### ISSUE TITLE
Graph Widget? Plot Widget?

### ISSUE THREAD

**Issue by ccoupe** [Shoes3/shoes3#264]
Does anyone need a challenge (  @passenger94 ?).  Long ago I did some some time series analysis foolery in Java and [wrote a "widget"](https://gist.github.com/ccoupe/5bde546796455b55ecd980b72b63e380) that would graph multiple time series data, auto scale the content to the window size, zoom in/out and move the zoomed graph left or right.  

Hard core data analysis folks would just call out to gnuplot but gnuplot is huge and not really something we want to include in Shoes.  Still, some cross platform visualization capability might be useful. I imagine, without thinking much on my part, that It would be Shoes widget (canvas) and we'd use mostly Cairo to draw and C/Ruby to get values out of Ruby arrays. Could it be done with pure Shoes/Ruby? Probably.  Performance? I don't know - the test data sets I have are 5K of observations and they drew surprisingly quick on a Windows 98 laptop in Java 1.02,

**Comment by ccoupe** [comment]
Wow! That old Java code still works and just as painful to setup. Screen shot of graphing ^SPX (S&P 500) and ^IRX.R (3 month US Treasury Bill rate). The question is how to do it in Shoes and not what the plot tells you. 
![graph-example](https://cloud.githubusercontent.com/assets/222691/17125559/b9bb8262-52b1-11e6-9fe6-03d14d402ca6.png)

**Comment by ccoupe** [comment]
[csv test data ](http://walkabout.mvmanila.com/public/share/test-csv.tgz
)

**Comment by passenger94** [comment]
> foolery in Java  

sounds like a pleonasm  :-D    

haha, thanks for challenge :-), will look at it, if only for curiosity, i was interested in this when looking at the profiler thing and ways to get visual feedback
lack of time lately, some opportunities maybe next month

**Comment by dredknight** [comment]
this looks really awesome!
I  have to get together and finish that popup thing and I can hop over and check this too :)

**Comment by ccoupe** [comment]
Still thinking out loud about this.  It should be easy enough to convert that java code into a  shoes.widget in pure ruby. Looking at the java, the DataSeries class is actually the bigger problem because I know it's sordid secrets - it has no way to represent 'no data' or 'bad data' for 'this observation'  and it assumes all DataSeries have the same date/time starting point and ending point and number of observations.  In order to be part of Shoes, the drawing code has to be in lib/shoes/ and while there are reasons to have DataSeries be a Gem, there are good (Loose Shoes symlink) reasons to put it into lib/shoes/DataSeries/

I've uploaded some Stock Market data files (csv format) to [here](http://walkabout.mvmanila.com/public/share/dataseries/) - the contents are obvious. There is a long ago, mostly useless reason they  start with '^' but it's what I have.

I've created branch 'graph' which is where this experiment should live.

**Comment by ccoupe** [comment]
Pure Shoes/ruby - test of concept. It works. [Of course it's missing all kinds of things.](https://gist.github.com/ccoupe/90524fab643c7a8eb442b7791ec3a06b) 
![pure_ruby_graph](https://cloud.githubusercontent.com/assets/222691/17468137/227b72d2-5ce2-11e6-921f-488596af430d.png)

**Comment by passenger94** [comment]
i say YES !!!
:high_heel:  :high_heel:  :high_heel:  :high_brightness:

**Comment by ccoupe** [comment]
So it's now a Shoes::Widget and it's  time to add the axes and labels. OOPS! Shoes doesn't have methods for drawing text at x,y. Time to move c/obj-c?  Maybe Shoes should have an Art Shape (like oval, star, etc ) named Art_String ???  so art_string {x:, y:, font:, fontsize:,color:,  length: , text:}. I don't know how hard that is to do but it has some appeal and might be useful for other folks for other applications. Probably easier to do  than creating a new c/obj-c graph widget so I like that. It might be useful text on top a image (with transforms already available like rotate and skew). Opinions?

**Comment by ccoupe** [comment]
Considering that I'm just converting my java code, it would be conceptually easier to create a widget,
similar to svg which I understand a lot better than Shape manipulation code (#266). It would have better performance for large data sets but not nearly as flexible for future enhancements. It's possible to do _both_ a native widget and a Shoes::Widget version.

I think I'll call the native widget 'plot' unless someone has a better name. API is pretty simplistic so far: 
`pl = plot {width:,  height: ,  [  {options_for_first} , {options_for_second} ]`} 
and `pl.add{{options_for_third})`  and `pl.redraw` in case someone is appending real time data. The option names: and defaults to be determined later but :data => [array of floatables]`, :name => "string", :minv => float, :maxv => float are the minimum.

**Comment by ccoupe** [comment]
API alert - some options belong to plot or chart (title and caption, width and height) and some belong to the data (like name, display_name, values and many more);  The plot/chart behaves differently depending on how many arrays it's supposed to display. So, it's a two step creation process in Shoes `plot width, height, {options)` and plot.add {many_hash_args}

**Comment by ccoupe** [comment]
Doesn't suck too much it you don't look. Fixable.

**Comment by ccoupe** [comment]
Getting better. So many things to do under the covers.
![better-plot](https://cloud.githubusercontent.com/assets/222691/17843130/a1bdc99a-67e7-11e6-96fc-7cc2a4ba1056.png)

The potential api keeps expanding with all the options the users might like ( turning of the grid lines for example)

**Comment by ccoupe** [comment]
I think this might be workable if you don't expect to have gnuplot like options in Shoes. The screen shot is  using a weird font "UnPilgi" (see graph.rb) because it odd ball. default is helvetica/
![pretty-good](https://cloud.githubusercontent.com/assets/222691/17882338/2b6eaea2-68c9-11e6-81a6-11b31654275e.

Much could be done and a few things _have_ to be done before I'm willing to merge this branch into the master.

**Comment by ccoupe** [comment]
![plot-nubs](https://cloud.githubusercontent.com/assets/222691/18192452/8afc0fd4-7091-11e6-8086-b4fdbe66a35a.png)

[updated](https://github.com/Shoes3/shoes3/wiki/Plot-Widget)

**Comment by passenger94** [comment]
synched and expert-graph.rb works fine on my system :+1:

**Comment by ccoupe** [comment]
Just committed  the latest. Two test programs  graph.rb and grtest.rb at top dir. 2nd one depend on a big csv data file - their are gc/cairo-destroy problem with very large collections of data.

**Comment by passenger94** [comment]
Woa ! I'm pretty sure plot.c is doing coffee if asked politely, !!!
will find the appropriate function tomorrow, bye :-)

**Comment by ccoupe** [comment]
Make coffee?  Why not? It's only a matter of time until someone asks for bar charts and pie charts and and ....  Actually, the #266 text shape would allow doing everything in ruby and a lot easier to code.

I really want to use pangocairo layout instead of the cairo toy interface but there are so many other things that need to be done first.

**Comment by ccoupe** [comment]
Can we more than one plot? 
![plottly-widgets](https://cloud.githubusercontent.com/assets/222691/18465168/90b3bc8c-7952-11e6-8fe6-371816f3fb42.png)

It's kind of widgety.

**Comment by passenger94** [comment]
"Bang for the buck" 
That's what i'm talking about ! :-D

**Comment by ccoupe** [comment]
There's nearly unlimited things that could be added. Visually oriented people might want so set the background color: plot  {background: "cornsilk")
![plot_background](https://cloud.githubusercontent.com/assets/222691/18503030/015731f6-7a18-11e6-84a6-414d7dd4c6af.png)

How hard would it be to draw a Column chart?
![column1](https://cloud.githubusercontent.com/assets/222691/18503797/1a20ed0e-7a1c-11e6-8045-9b465d258115.png)

That needs some tuning.

**Comment by passenger94** [comment]
![applause](https://cloud.githubusercontent.com/assets/10730930/18514559/530b4d50-7a92-11e6-83fc-c6a0fa5953a0.gif)

**Comment by ccoupe** [comment]
Nearing completion of the bar chart (I hope).
![bar-graph](https://cloud.githubusercontent.com/assets/222691/18611918/6520b49a-7d07-11e6-98a2-cd91396b1c56.png)

There is a problem - drawing x axis labels can't be set to dark gray.

**Comment by ccoupe** [comment]
So a scatter chart is like a line chart without the connecting lines. How hard could that be?

![scatter1](https://cloud.githubusercontent.com/assets/222691/18623526/89ac9f50-7dfb-11e6-9558-126121231fdb.png)

Actually, a little easier than I thought and a little harder because it exposed some glossed over drawing issues.

**Comment by ccoupe** [comment]
Turns out that is not what a scatter chart is according to my ancient papyrus Excel manual.

**Comment by ccoupe** [comment]
An _almost good_ scatter plot. I wish I knew what I was doing.
![almost-good-scatter](https://cloud.githubusercontent.com/assets/222691/18805602/7539d164-81cf-11e6-89a0-67363c3a9c5f.png)

**Comment by ccoupe** [comment]
On the shiny side of life:
![pie-begin](https://cloud.githubusercontent.com/assets/222691/18901361/7d1b88a6-8507-11e6-96a5-1dea40fdddd8.png)

On the dull side of life, there are some odd problems with color choice for the wedges. The legend is correct.

**Comment by dredknight** [comment]
Dang this looks awesome... I already know how I am going to use it :D.
Great job!!!

**Comment by BackOrder** [comment]
Are we going to have some [radar chart](https://en.wikipedia.org/wiki/Radar_chart)? Also perhaps some way to easily extend the feature with new graphs?

![image](https://cloud.githubusercontent.com/assets/8844578/18954329/88b80ed6-8620-11e6-9807-902e987febbf.png)

![image](https://cloud.githubusercontent.com/assets/8844578/18954337/8f933532-8620-11e6-93d2-016133e42da4.png)


![image](https://cloud.githubusercontent.com/assets/8844578/18954327/7e2c68ea-8620-11e6-951e-95275baae8b4.png)

**Comment by ccoupe** [comment]
There is no user extensible charting for this feature (#264). It's all done in C and cairo. If I get pie charts working I'll probably pause there.  There are many things missing in the existing charts and I suspect errors too.  And there are testing problems. One could spend a lifetime recreating charting functionality in Shoes when better standalone programs exist. OpenOffice/LibreOffice are free, better and easier to work with.  Large data sets or complex analytics seem to be done in R+gnuplot or Matlib or other programs that Shoes shouldn't attempt to be (IMO). 

If #266 was implemented then you have everything needed to implement your chart in Ruby in a Shoes user Widget (same painful problems as C though).

**Comment by ccoupe** [comment]
One interesting idea for extensibility, contemplate is [rcairo](https://github.com/rcairo/rcairo) (it's a gem so we can include it in Shoes in several ways). Perhaps a new widget? Call it raw_widget,  for example
```ruby
Shoes.app do
  raw_widget 600, 400, do 
     require 'rcairo'
     #ruby_cairo drawing code here
  end
end
```
That Shoes widget just calls the user's block at draw time and Shoes manages the rectangle placement in the overall layout. If you can figure out cairo & rcairo, you could draw damn near anything you want.  Cairo is not friendly for beginners but it can also be easier than managing  shoes shapes (#266).  Just a thought that I wish I'd thought of earlier.

**Comment by BackOrder** [comment]
It might be possible to write an overriding drawing method, or some kind of call back, that is attaching to your existing C code. Normally the input data is always the same. The difference is the representation on screen. The callback would be on Ruby level.

**Comment by ccoupe** [comment]
@BackOrder, sadly that would require a Shoes api (new methods) into the chart drawing code for the user to call and so far there is little in common in all the chart types that could be exposed to Shoes/Ruby and woe be those who call it incorrectly or out of order  since it runs deep inside  gtk/cocoa event draw code you may not  get ruby or toolkit exceptions (I know this for a fact); 

There is  some [abandoned python code ](https://bitbucket.org/lgs/pycha/src) that I'm slowly converting to C for the pie chat that  might convince you that it a charting framework  can be done in pure Ruby/Shoes if some one want to do that (and #266 _or_ the rcairo thing), It has something called a ringchart.py which might be close to your radar chart. Or not.

**Comment by ccoupe** [comment]
Nearing completion. Labels around the pie.
![pie-good-enough](https://cloud.githubusercontent.com/assets/222691/19018129/3262cb3a-8813-11e6-8a21-e0c8ef063ff1.png)

Raw values are the default labels. % is an option

**Comment by ccoupe** [comment]
Optional box around the legend - see update [wiki](https://github.com/Shoes3/shoes3/wiki/Plot-Widget) page for an example of how that looks. 

My search for radar chart producing code to crib from turns up very little. In the [pycha ](https://bitbucket.org/lgs/pycha/overview) there is a radial.py chart that might be close (or not) but its only used in his 'chavier' app and that has a startup bug in my python (2.7) and my python skills are low. Help? (it uses pygtk and pycairo so that my prevent some Windows or OSX help - don't know).

I think I get the idea:  Using the download example data shown above, there would be 8 lines drawn from the center point (of equal angle - the length of the line would be scaled by it's % (above). Draw a litlte dot there and perhaps the value label. - Then draw lines connecting each of those dots. Is that correct? A helpful someone could even do the drawing the line and math  in Shoes/Ruby and then I'd really have good example.

**Comment by ccoupe** [comment]
LibreOffice has a Net chart type that behaves almost like I describe above - the length of the radial axis lines scale to the largest value in the data set - I mistakenly said % of sum.  Here's a Net using download data above ` [15, 5, 4, 4, 3, 3, 1, 1]` 
![libreoffice-net](https://cloud.githubusercontent.com/assets/222691/19061782/d01bc8d8-89b0-11e6-81de-8b8da9737ac4.png)

Not useful for me for this data but the important point is whether that is what @BackOrder is thinking?

**Comment by BackOrder** [comment]
Something along this line... [Radar charts](https://en.wikipedia.org/wiki/Radar_chart) has many alternative names. Let me show you a real case scenario that I wrote before... [Statistics Feature for Humintell Products](https://fr.scribd.com/document/28961133/Statistics-for-Humintell-Products). It should give you a better idea.

**Comment by ccoupe** [comment]
Thanks, @BackOrder .  They are similar charts (LibreOffice _Net_ and what you used) I've created a gr6.rb test case based on the Advertising sample you provided -- assuming I can convert that radial.py from pycha. I can also see why some folks consider radar as mis-leading eye candy. Not my fight. Probably left brain or right-brain related.

**Comment by BackOrder** [comment]
Charts are charts. One could argue that the people are those whom are misleading rather than the charts.

The way radar chart is used in the document aforementioned is rather to visually see someone's progress, or area of improvements, at key point in time (pre-training, during training, post-training).

**Comment by ccoupe** [comment]
> One could argue that the people are those whom are misleading rather than the charts.

Too true. I've seen bad, good and ineffective charts. I've created some of each myself.  Another reason I don't want to get  too complex in Shoes plots.

**Comment by ccoupe** [comment]
I think this is a reasonable target for Shoes - 
![radar-goal](https://cloud.githubusercontent.com/assets/222691/19098860/1e91c3a6-8a6e-11e6-91b1-871f701cc563.png)

Not nearly as pretty and LibreOffice goes counter clockwise abd the msft package goes clockwise. Minor point if you think about it.

**Comment by ccoupe** [comment]
I managed to get the pycha app, _chavier_ working.  Too bad it's abandoned, sigh - its useful with bugs of course. I was able to entering the advertising sample (Test/plot/gr6.rb value) and it produced this for it's radial chart type:
![chavier-radial-1](https://cloud.githubusercontent.com/assets/222691/19140031/aa275074-8b46-11e6-8587-27e8bf4af3d5.png)

That's damn close to what I was looking for which means I can probably  use the radial.py python code and convert it to C/cairo, In case you want fix your copy of chavier/app.py, around app.py:96, make it look like
```python
        legend=dict(
            opacity=float,
            borderColor=str,
            borderWidth=int,
            hide=bool,
            legendFont=str,
            legendFontSize=int,
            position=dict(
...

**Comment by BackOrder** [comment]
```ruby
@grf.add num_obs: @values1.size, values: @values1, xobs: @xobs,
      name: "Year 1", minv: 0, maxv: 200, color: "dodgerblue",
      nubs: "dot", strokewidth: 1
```

Why do we have to send the size when an array is sent to values:? 
Normally the number of items would be calculated internally.

**Comment by ccoupe** [comment]
Good Question. I could not find a documented C level api that returns the length of an array. It also complicates things at the C level. I'd be happy to be pointed to the C call for it but I couldn't find it.

**Comment by BackOrder** [comment]
Just a wild guess... ```rb_ary_size()```

**Comment by ccoupe** [comment]
Doc - https://github.com/ruby/ruby/blob/6a69ab937c7e6584d311cd444615cc1cd5b5499b/doc/extension.rdoc . No doubt there is an undocumented way otherwise ruby couldn't work.

**Comment by ccoupe** [comment]
macro RARRAY_LEN(value) seems to work.  It's existence is hinted at if you know about RSTRING_LEN and read the doc _very_ carefully. Excellent. num_obs: is now ignored.  Thanks.

**Comment by BackOrder** [comment]
Very good job.

I'd also like to recommend changing **xobs** to **labels**, which is more comprehensible for anyone whom ever use charts. **minv** and **maxv** could be just **min** and **max**.

What is **nubs** ?

**Comment by ccoupe** [comment]
Reasonable suggestions and easy to do.  
>What is nubs ?
Nubs are those little boxs or dots or diamonds drawn in line and scatter charts. I don't know what they should be called.  I think I documented them on the wiki article.  Since I haven't written the shoes manual yet, now is good time to be changing these things.

**Comment by BackOrder** [comment]
Google calls them [Points](https://developers.google.com/chart/interactive/docs/points), isn't it?

Alternatives: LineStyle, Style, LineType, Type, PointShape, Shape, etc.

Point, Style or Shape could do it. Short and simple.

How does LibreOffice call them?

**Comment by ccoupe** [comment]
LibreOffice calls them Points. Works for me.

**Comment by BackOrder** [comment]
Oh! Would **long_name** be the **legend**?

**Comment by ccoupe** [comment]
Please see https://github.com/Shoes3/shoes3/wiki/Plot-Widget. Yes, name and long_name are displayed  in the legend area. long_name is optional, name is not. If specified, it's used in the legend.

missing controls what to show when there is missing data (nil) in data series. Again, see that wiki article.

**Comment by BackOrder** [comment]
Then I'd recommend **description** or **desc** instead of long_name. The latter doesn't say much about what it is. How about changing **missing** to **default** as in _default value_?

The plot API is rich in features. You did a really good job on its implementation.

**Comment by ccoupe** [comment]
Again, **desc** and **default** are reasonable suggestions.  I have no objections. I just made up names that meant something to me so I could write some code.

Knowing when to stop adding features is an issue. So many _things_ things could be done.

**Comment by ccoupe** [comment]
A comment about the last commit.  I've started the work on two things.  _Eventually_, in the `shoes_plot` struct (canvas.h),  I'll remove all those Arrays (maxvs, minvs, values, names, long_names, xobs, strokes,  nubs,  color) - replaced be an Array of `chart_series` objects). chart_series contains those things flattened of course and uses @BackOrder 's desired names (at the C and Shoes level)

The chart_series objects are created without  the user's knowledge from the arguments to plot.add. There is untested code for Shoes users to create a chart_series themselves and add that object to a plot instead of a hash of arguments. 

There is code handle both the old way and the new way as each chart type is converted to the new way.

**Comment by BackOrder** [comment]
By the way, why is it called **chart_series** instead of just simply **chart**? Simplicity is always key but you may have good reasons for the name.

**Comment by ccoupe** [comment]
Although plot_series or plot_data or something like that is a more logical name. My hope is somebody will step forward and write a graph library /gem in Shoes/Ruby (assuming #266) and they'll probably name it Chart or Graph.  If they have lots of data residing in files or sql db query results, instead of re-typing values and labels into their Shoes script, then they need an Object of some_name to hold that data and it should be usable in both the built in Shoes and their shiny new library or gem or script. Psuedo code:
```ruby
class DataSeries << chart_series do
  require 'sqlitite'
  def initialize (sqlquery_string)
    ....
   super(values, labels, min, max...)
  # add my special things
  end
end
```
That  object could be used by both the Built in Shoes plot types and anything the (future) #266 user creates. 

I'll note that most of the world  have never seen or programmed a mechanical plotter and debugged a FORTRAN/C/driver and the users program. Life is much better now days. Much better. Hell, you drive a 3D pringer with Raspberry PI and and never know about C variable names in the library.

The intent was that this should be subclassed for #266 but usab

**Comment by BackOrder** [comment]
This is very interesting. I'm liking the thoughtfulness of your approach. It does seem like a wrapper class more than a base class. Your chart_series and plot_series already do the deed. So they should really name their class like **SQLiteChart**.

Shoes is meant to be intuitive and easy to use. Wouldn't those be arguments to name them **chart** and **plot**?

My concern is really Shoes end users. Their programming background may be limited. They use Shoes for convenience and easy to learn and use. Long names, complex names, meaningless names are all kind of things to throw them off their chairs and give up.

**Comment by ccoupe** [comment]
Evidence suggests Shoes doesn't have a lot of users **or** the product is nearly perfect and they have few complaints. People occasionally  contact me via email and everyone of them has been pretty knowledgeable. If you're clever enough to figure out sql in ruby (ugly) or contemplating writing your graph library or widget using Shoes shapes then you'll be able to handle chart_series as a name. A decent argument can be made for calling it plot_series but I'm not convinced it's really going to matter. 

I do intend to add some methods to the chart_series class. I also want to allow a &blk arg (proc) that will be passed some args: index of the label, and how many characters the current layout will allow. Uses can size down the return string any way that works for them, or just let the labels overlap visually. Not written yet and I suspect no-one will use it when it is written.

The chart_series will be documented in the Advanced section of the manual with just a few mentions in the Elements->plot section. 

Remember, one of the goals of Shoes  3.3  was to expand the range of problems Shoes could address for more skilled developers.  New terminal/console window (optional ) for old school `puts` from things like mini_test (not a beginner tool) and the  Profiler (not a beginner tool) and better command line support. Merge packaging is for hard core professional developers.

The  Simple Shoes Vision no longer exists. I suspect it when _why left the building. Ruby is too complex for something simple. If someone cared about new users and tutorials, they would maintain Hackety Hack. Crickets. Is Shoes easier than pygtk ? Yes. As powerful? No.

**Comment by dredknight** [comment]
@ccoupe facts are on point. When one is a newbie in coding he does not search for easy tool to do stuff. He just starts with something popular because when you don't know anything you start with what is modern/fashionable.

I love Shoes because of its syntax simplicity which does not make it limited when it comes to building complex stuff.

**Comment by BackOrder** [comment]
@ccoupe It seems that you feel much strongly about the issue than I do. However, I'd like to reply to what you wrote for your consideration.

> Evidence suggests Shoes doesn't have a lot of users or the product is nearly perfect and they have few complaints. People occasionally contact me via email and everyone of them has been pretty knowledgeable. 

It is true now. It might be true later on. It might or might now always be true. On the other side, keeping the syntax of new features friendly will always leave the door open for newcomers. Also, remember, consistency is important. We are talking about a language where **paragraph** is shorten to **para**. 

> If you're clever enough to figure out sql in ruby (ugly) or contemplating writing your graph library or widget using Shoes shapes then you'll be able to handle chart_series as a name. A decent argument can be made for calling it plot_series but I'm not convinced it's really going to matter.

Ah! You are missing my point. I am talking about the rest who would not use SQL. They would use chart_series or plot_series directly. My understanding is that they draw charts with data provided by the user in a self-sufficient manner.

Hypothetical newbie scenario:

- Hey, I'd like to draw a pie chart...
- Let's see what Shoes has for that... **chart**? Sounds good
- Alright, where do I enter the title... oh yeah... the labels? right... data? done.
- Type, type, type... run... wow, it works!

Just like when someone wants to draw a line, circle, svg, etc. The Shoes commands are appropriately corresponding to what someone would one to do. No fancy thing.

> I do intend to add some methods to the chart_series class. I also want to allow a &blk arg (proc) that will be passed some args: index of the label, and how many characters the current layout will allow. Uses can size down the return string any way that works for them, or just let the labels overlap visually. Not written yet and I suspect no-one will use it when it is written.

Amazing!

> The Simple Shoes Vision no longer exists. I suspect it when _why left the building. Ruby is too complex for something simple. If someone cared about new users and tutorials, they would maintain Hackety Hack. Crickets. Is Shoes easier than pygtk ? Yes. As powerful? No.

Not entirely true but not entirely false. Consistency should however be taken in consideration. We might not be able to redo the whole underneath model as in Shoes4 but we still can keep things in a straightforward manner.

**Comment by ccoupe** [comment]
As we often do, we seem to be talking past each other until we figure it out.  Remember that chart creation is a two step process - create the widget (chart type) and then add data/labels to it. For line, bar, and timeseries, you can add multiple collections of '_Thing X_' to the chart. _Thing X_ also contains important things like the color you want and the min/max used to scale that part of the chart.

 I really don't care what the Shoes object is named, I think it needs to exist and it does exist because the C was getting unmanageable and that wouldn't get better for new chart types or future enhancements to Thing X.   From Tests/plot/gr3.rb
```ruby
    @grf2.add values: @values1, labels: @x_axis1,
       name: "Bar", min: 0, max:  30, desc: "foobar Yy", color: "crimson",
       points: true, strokewidth: 12
    cs2 = chart_series values: @values2, labels: @x_axis2,
       name: "Tab", min: 50, max: 230, desc: "BarTab", color: "green",
       points: true, strokewidth: 6
    @grf2.add cs2
```
Both @grf.add 'hash' or 'chart_series'  work.  User's preference. The 'hash' version just creates the chart_series for you behind the scenes.  Which one is the simplest? Minor differences, IMO and we have both, and always will. .  Which one can be sub classed and re-purposed for other duty should someone who is not newbie prefer?  - hash args in a Ruby program they cant change by a running script or an object they can build on? 

Currently _Thing X_ is called 'chart_series' at the C and Ruby levels. In my wandering around the interwebs I haven't found a common name for what _Thing X_ is. I also have seen may FOSS windows or OSX programs that draws simple charts, simply. Niche?  I do care about doing the best thing given the situation we find ourselves in. We all do. Actually  I prefer data_series and shoes_plot_data_series  but that feels presumptive and invites a name clash with some ruby gem somewhere.

**Comment by BackOrder** [comment]
@ccoupe thanks for explaining. Hash is simpler. ;)

By the way, does color: suport rgb() values? It should be noted that other Shoes methods support Shoes colornames, such as **background**, which does not use quotes as in Test/plot/gr3.rb. Consistency is important.

Hopefully you understand that I am not nitpicking. The chart feature is absolutely fantastic. I remember wanting to get some Java charting in the past but the only decent packages were available for sales. So it's refreshing to get it for Shoes.

**Comment by passenger94** [comment]
@BackOrder, yes i do agree, C functions should check for a ruby color (rgb or named) in the hash given to plot related methods (like texts, shapes, etc ... do for stroke, fill and those you mentioned, background, border ...). 
The whole plot extravaganza is shiny enough now to spend time on killer "details" :-)

**Comment by ccoupe** [comment]
I welcome help and you are helping:  Turns out the plot color can easily be handled as you suggest.  Bonus points because I get to delete some ugly code. I wasn't fond of the string myself.  Fixed.

Another annoyance is having to use strings for "dot" and "rect" - I'd prefer symbols :dot and :rect . Opinions? 

The radar/spider chart ineed a lot help if someone has some extra time.

**Comment by BackOrder** [comment]
Great!

> Another annoyance is having to use strings for "dot" and "rect" - I'd prefer symbols :dot and :rect . Opinions?

Can't we have similar thing as Shoes colours? That would end up with **dot** and **rect** directly (without quotes or colon). Otherwise symbols would be fine as far as Ruby goes but it's always a bit strange to use with named parameters, such as **points: :dot** (a lot of colons if you notice). 

> The radar/spider chart ineed a lot help if someone has some extra time.

What's the TODO list on this one?

**Comment by ccoupe** [comment]
For better or worse, all of the color names are real Shoes objects in the Shoes name space.  While we could do something similar and create Shoes classes for dot, box, rect, and circle,  I am not comfortable that - it  could to break existing scripts that have those class  names.  At the C level, I'm not happy to pollute Ruby space with the :symbols either but since "Shoes is not Ruby it only looks like it",  symbols seem a better choice at the moment.  (assuming I do anything)

> What's the TODO list on this one? 

Huge. I don't know a sin() from a cos(), radians from degrees, and when to use one vs the other.  The code in there is wrong (converted from pycha). One of the magazines I read (Car and Driver) presents their  results in a spider graphs (they also get complaints about them - but that's not our problem to solve). Radial axis can run from small number near the center to larger number on the edge. Or larger numbers at the center and smaller at the edge - depending on how they define goodness.  - a reversal on out min/max  - The code in there might handle that, might not - I suspect  not. Then there is issue of labels. One of the charts in your example way above this comment appears to use axis in 0->100,000 but the labels are 50, 70, 90. Another reason for the &blk the chart_series/hash/Thing_X, add. 
center (a reverse of what my current min/max likes)

**Comment by BackOrder** [comment]
> re: namespace

This is an important question. Creating additional Shoes objects would promote consistency and better syntax but your argument that it might break existing code is absolutely valid. It may not have big impact or may be expected from upgrading process but it's still important enough to consider before moving forward with such approach.

Ruby symbols would be an alternative and we could depreciate Shoes objects (such as colour names) via a warning in Shoes console to once again promote consistency. The depreciation would include support of colour name as symbols and we would phase out Shoes objects some future release.

This is something to think about because Shoes has been having increasingly new features since you took over the development. It comes to reason we need to have some kind of guideline.

> re: math

Ruby Math (and most programming API) only support radians. One radian equals 1/2π. Degrees are something human enjoys and goes from 0 to 360 degrees for a full circle; exceeding values results in circling again, negative values for counterclockwise. 

What you need to know is that 2π radian equals full circle (360 degrees). You use this reference to know how to split equally. You have X values on a radar chart? 2π / X.

Something along this line... not sure whether it is helpful to you...

**Comment by ccoupe** [comment]
For now, I'm going to leave points and a few others as strings args until I think harder about symbols and learn how to use them at the C level. I think I know and what I think is ok-maybe. 

>What you need to know is that 2π radian equals full circle (360 degrees). You use this reference to know how to split equally. You have X values on a radar chart? 2π / X.

That bit of knowledge helps. I 'll review the  the pycha radar to C/cairo conversion with that extra info.

**Comment by ccoupe** [comment]
One use case for chart_series (aka Thing_X):  Imagine you want to monitor the temperature of a probe in a BBQ smoker - when its running between set points of 225F and 240F you draw a black line/datapoint on the graph. If it drops below 225F then draw a blue line and if rises above 240F then draw a red line. Imagine it spikes up to 280 and you need rescale the y axis too. You'll need getters and setters for all the hash arguments for chart_series (plot.add) but if you don't have a backing object then those would be getters and setters on the plot object and each of them would require an index or hash arg to pick which series to modify. Unclean in C and unclean in Shoes.

See Test/plots/cstest.rb - it appends (label, value) but it can also replace existing (labels, values) - just change that 7 to a 3 for example. Also note in gr4.rb that is uses the return value of a plot.add (currently named chart_series).

**Comment by BackOrder** [comment]
So Shoes charting allow realtime data and other modifications?

**Comment by ccoupe** [comment]
> So Shoes charting allow realtime data and other modifications?

One of my design goals. Fundamental.

**Comment by BackOrder** [comment]
I'm looking forward to it. :D

**Comment by ccoupe** [comment]
Already there in gr1.rb which adds random data points every 5 seconds.

**Comment by ccoupe** [comment]
[New beta available](http://walkabout.mvmanila.com/public/shoes/) . Seems to mostly work Windows for graph tests but you need the graph test script which you only get from github sync. Perhaps my test scripts need to be in samples? expert-plot1.rb ...expert-plot6.rb ? Ick.

**Comment by ccoupe** [comment]
Who knew? [Gruff](https://github.com/topfunky/gruff) is a ruby (rails gem) that draws all kinds of charts, including spider, using rmagick as the backend but at the Ruby level the code is pretty similar to what C/Cairo uses. Anyone attempting #264 charting package  might want to start with this.

**Comment by BackOrder** [comment]
It would have been nice but Gruff doesn't seem maintained anymore. The last commit is 5 months old and issues are dating in years.

Furthermore, rmagick as a backend might not be suitable for realtime graphs.

**Comment by ccoupe** [comment]
Gruff is source code to copy and modify ourselves. It's not too old and it does work within it's limits and they have test data, scripts and test output. It's hard to install (even on ubuntu and probably very difficult for windows). I would never use it as a gem to include in Shoes.

 The algorithms inside are the value to me (and the test's,  so I know that it should look like).  I doubt anyone will modify the Ruby code to use Shoes shape drawing code in place of rmagick , but they could (#266).

**Comment by ccoupe** [comment]
We have radial axis.
![radar-1](https://cloud.githubusercontent.com/assets/222691/19834812/420d8312-9e35-11e6-9a48-1645edb87db5.png)

**Comment by BackOrder** [comment]
Good. I'd like to recommend a 90 degrees rotation to match other radar graphs samples we shown here (e.g. LibreOffice). You will certainly notice the starting point on your graph is on right-middle quadrant. LibreOffice et co. would normally start on top-middle quadrant.

**Comment by ccoupe** [comment]
I was working from the simpler spider in gruff which for whatever reason draws it with that start position.  What we really want is Gruff's  'net' chart, which is a bit more complex but easier to comprehend than the pycha code.

**Comment by ccoupe** [comment]
This issue is getting a little slow to load and it won't get faster.  I've opened #288 for radar/spider issues.

**Comment by BackOrder** [comment]
I am not a huge fan of **pie_percent** on [L019](https://github.com/Shoes3/shoes3/blob/e88caec2aa6a91d0c3c06e9c88b2fcf463373647/shoes/plot/plot.c#L109). How about just **percent**?

**Comment by ccoupe** [comment]
Closed in Shoes 3.3.2
